src/qibolab/instruments/qm/config/pulses.py

```diff
@@ -0,0 +1,137 @@
+from dataclasses import dataclass, field
+from typing import Union
+
+import numpy as np
+
+from qibolab.pulses import Pulse, Rectangular
+
+SAMPLING_RATE = 1
+"""Sampling rate of Quantum Machines OPX in GSps."""
+
+__all__ = [
+    "operation",
+    "Waveform",
+    "waveforms_from_pulse",
+    "integration_weights",
+    "QmPulse",
+    "QmAcquisition",
+]
+
+
+def operation(pulse):
+    """Generate operation name in QM ``config`` for the given pulse."""
+    return str(hash(pulse))
+
+
+def _normalize_phase(phase: float):
+    return (phase % (2 * np.pi)) / (2 * np.pi)
+
+
+@dataclass(frozen=True)
+class ConstantWaveform:
+    sample: float
+    type: str = "constant"
+
+    @classmethod
+    def from_pulse(cls, pulse: Pulse):
+        phase = _normalize_phase(pulse.relative_phase)
+        return {
+            "I": cls(pulse.amplitude * np.cos(phase)),
+            "Q": cls(pulse.amplitude * np.sin(phase)),
+        }
+
+
+@dataclass(frozen=True)
+class ArbitraryWaveform:
+    samples: list[float]
+    type: str = "arbitrary"
+
+    @classmethod
+    def from_pulse(cls, pulse: Pulse):
+        phase = _normalize_phase(pulse.relative_phase)
+        samples_i = pulse.i(SAMPLING_RATE)
+        samples_q = pulse.q(SAMPLING_RATE)
+        return {
+            "I": cls(samples_i * np.cos(phase) - samples_q * np.sin(phase)),
+            "Q": cls(samples_i * np.sin(phase) + samples_q * np.cos(phase)),
+        }
```

Comment:
I guess we can easily lift this to the general pulse library. On the other hand, we could also wait to see if this is actually useful for other instruments before doing so.

Reply:
> On the other hand, we could also wait to see if this is actually useful for other instruments before doing so.

I partially agree with you. Though, this is *potentially* useful, and the price of lifting is almost negligible (in terms of code and readability). Thus, in general, I'd work in advance to prevent duplication.

In any case, it's truly not that relevant.

Reply:
Implemented in 22d599307b44d689774b262d13e8ade07f2cbc37 and tested that it doesn't break anything. I will now merge this PR, if you like to change something in that implementation, we can do so in a different PR.